### PR TITLE
migrate firehose to ASF

### PR DESCRIPTION
**Pull request by alexrashed**
This PR migrates the existing Flask based Firehose API to our new AWS Service Framework.
Besides adding tagging support, this PR does not introduce any new features.
However, it removes a lot of the boilerplate request/response handling code, cleans up a lot of the dict handling with the generated ASF types, and builds the foundation for further enhancing this API in the future (f.e. with the upcoming OpenSearch integration - https://github.com/localstack/localstack/issues/5357).

### CODE REVIEW

File: localstack/services/firehose/provider.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+            destinations.append(
+                DestinationDescription(
+                    DestinationId=short_uid(),
+                    ElasticsearchDestinationDescription=convert_es_config_to_desc(
+                        elasticsearch_destination_configuration
+                    ),
+                )
+            )
+        if amazonopensearchservice_destination_configuration:
+            destinations.append(
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:168]: To me this implies that a Firehose can be configured to deliver to both an ElasticSearch and OpenSearch destination simultaneously. Is that accurate?

File: localstack/services/firehose/provider.py [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+        if http_endpoint_destination_configuration:
+            destinations.append(
+                DestinationDescription(
+                    DestinationId=short_uid(),
+                    HttpEndpointDestinationDescription=convert_http_config_to_desc(
+                        http_endpoint_destination_configuration
+                    ),
+                )
+            )
+
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:197]: It looks like not all of the possible destinations are accounted for here in this `destinations.append()` logic (e.g. Splunk, Redshift). If I'm following this correctly, it looks like you could invoke `create_delivery_stream()` with arguments for these non-implemented destinations, but the Firehose would make no attempt to deliver to them. Is that accurate? If so, would it be worth emitting some kind of warning in these cases?

File: localstack/services/firehose/provider.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+                    records = self._preprocess_records(processor, records)
+
+            if "ElasticsearchDestinationDescription" in destination:
+                es_dest = destination["ElasticsearchDestinationDescription"]
+                es_index = es_dest["IndexName"]
+                es_type = es_dest.get("TypeName")
+                es = connect_elasticsearch(
+                    endpoint=es_dest.get("ClusterEndpoint"), domain=es_dest.get("DomainARN")
+                )
+                # TODO support FailedDocumentsOnly as well
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:463]: Consider emitting a warning `FailedDocumentsOnly` is set

File: localstack/services/firehose/provider.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+                    try:
+                        es.create(index=es_index, doc_type=es_type, id=obj_id, body=body)
+                    except Exception as e:
+                        LOG.exception(f"Unable to put record to stream {delivery_stream_name}.")
+                        raise e
+            if "AmazonopensearchserviceDestinationDescription" in destination:
+                # TODO implement opensearch integration
+                # - Replace the elasticsearch client with the opensearch client
+                # - Then this most likely can be handled together with ElasticsearchDestinationDescription
+                # opensearch_dest = destination["AmazonopensearchserviceDestinationDescription"]
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:506]: 👍

File: localstack/services/firehose/provider.py [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+                try:
+                    requests.post(url, json=record_to_send, headers=headers)
+                except Exception as e:
+                    LOG.exception(f"Unable to put Firehose records to HTTP endpoint {url}.")
+                    raise e
+        return [
+            PutRecordBatchResponseEntry(RecordId=str(uuid.uuid4())) for _ in unprocessed_records
+        ]
+
+    def _add_missing_record_attributes(self, records: List[Dict]) -> None:
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:535]: is there a specific case you have in mind where these attributes would be missing or is this just for general safety/robustness?

File: localstack/services/firehose/provider.py [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+                if "data" in record:
+                    record["data"] = to_str(record["data"])
+                if "Data" in record:
+                    record["Data"] = to_str(record["Data"])
+            event = {"records": records}
+            event = to_bytes(json.dumps(event))
+            response = client.invoke(FunctionName=lambda_arn, Payload=event)
+            result = response.get("Payload").read()
+            result = json.loads(to_str(result))
+            records = result.get("records", []) if result else []
```
Comment (SwatsonCodes) [inline localstack/services/firehose/provider.py:576]: Does this mean that if the Lambda function errors out or returns some invalid response that records we attempted to process are silently dropped?

File: localstack/aws/api/firehose/__init__.py [hunk 7]
```diff
@@ -0,0 +1,1860 @@
...
+RedshiftRetryDurationInSeconds = int
+RetryDurationInSeconds = int
+RoleARN = str
+SizeInMBs = int
+SplunkRetryDurationInSeconds = int
+TagKey = str
+TagValue = str
+Username = str
+
+
```
Comment (SwatsonCodes) [inline localstack/aws/api/firehose/__init__.py:79]: Did you write out all the class definitions, or were they harvested from somewhere else? If it's the latter, could this be accomplished via some import?

**Comment by SwatsonCodes** [comment]
Put up a few questions and comments but overall this looks like a solid, straightforward foundation.
One last thought: it seems like the integration testing is pretty light. What do you think about creating a separate issue for adding more thorough testing? Firehose is a fairly complex service with lots of potential for corner cases so I think it could be worth investing time into a robust test suite - especially as ElasticSearch and OpenSearch inevitably diverge we'll want to catch errors proactively.

File: localstack/services/firehose/provider.py [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+        context: RequestContext,
+        delivery_stream_name: DeliveryStreamName,
+        tag_keys: TagKeyList,
+    ) -> UntagDeliveryStreamOutput:
+        delivery_stream_description = _get_description_or_raise_not_found(
+            context, delivery_stream_name
+        )
+        region = FirehoseBackend.get()
+        # The tagging service returns a dictionary with the given root name
+        region.tags.untag_resource(
```
Comment (whummer) [inline localstack/services/firehose/provider.py:349]: nit: Could be simplified to `FirehoseBackend.tags.untag_resource(..)`, given that `tags` is a static class member. (in fact, we could rename it to `TAGS`, to make this a bit more clear.)

**Review by whummer — APPROVED** [review]
Awesome changes - great to see another Flask app getting ripped out! 💯

File: localstack/services/firehose/provider.py (hunk 1 above)
Comment (alexrashed) [reply localstack/services/firehose/provider.py:168]: Yes, that's correct. :)

File: localstack/aws/api/firehose/__init__.py (hunk 7 above)
Comment (alexrashed) [reply localstack/aws/api/firehose/__init__.py:79]: The content of this file is generated with our scaffold (`localstack/aws/scaffold.py`).
So it is generated using our own code generation when we start implementing a new service.
However, the implementation is tightly coupled to the generated code (and it takes some time to parse the specs and generate the code). Therefore we decided to commit the generated code in the repo.
In the future we will have a "generated code" warning on top of each file and might introduce a bot which automatically tries to regenerate the code (to automatically update to non-breaking changes in the specs) and creates PRs.

File: localstack/services/firehose/provider.py (hunk 6 above)
Comment (alexrashed) [reply localstack/services/firehose/provider.py:576]: Yes, that's correct. However, I am not sure how the AWS service handles failures of processing lambdas (I just migrated the behavior as it was before).

File: localstack/services/firehose/provider.py (hunk 2 above)
Comment (alexrashed) [reply localstack/services/firehose/provider.py:197]: Correct. In LocalStack we currently only support those four destinations.
Good point to the warning, I'll one :)

File: localstack/services/firehose/provider.py (hunk 5 above)
Comment (alexrashed) [reply localstack/services/firehose/provider.py:535]: The method is a bit generically named, but essentially it currently only adds Kinesis metadata if it's missing.

File: localstack/services/firehose/provider.py (hunk 6 above)
Comment (SwatsonCodes) [reply localstack/services/firehose/provider.py:586]: I'm not sure either actually, I've never had to use the Lambda integration 🤔 I'm guessing the records just get put into S3 like in other failure cases with Firehose. Might want to add a warning here.

File: localstack/aws/api/firehose/__init__.py (hunk 7 above)
Comment (SwatsonCodes) [reply localstack/aws/api/firehose/__init__.py:79]: cool!

**Pull request merged**